Value range refinement using uni-variate expressions.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97964
* __->__ #97963
* #103611
* #102563

This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

ysiraichi committed:
Value range refinement using uni-variate expressions.

This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi commented:
Number of issued guards does not seem to change from [the one reported previously](https://github.com/pytorch/pytorch/pull/96616#issuecomment-1482547555).

One notable change is for: `attention_is_all_you_need_pytorch`. The guards were reduced due to mirroring the guard expression and re-running the value refinement procedure.

_Each number separated by a '/' represents the number of issued guards for a given graph._

| Benchmark                         | Guards (before)   | Guards (dedup)   | Guards (After)   |
|-----------------------------------|-------------------|------------------|------------------|
| Background_Matting                | 36/**69**         | 36/**67**        | 36/**61**        |
| LearningToPaint                   | 9/**52**          | 9/**43**         | 9/**21**         |
| attention_is_all_you_need_pytorch | 10/32             | 10/32            | 10/**24**            |
| dcgan                             | 9/**71**          | 9/**64**         | 9/**27**         |
| dlrm                              | 34/**48**         | 34/**42**        | 34/**36**        |
| drq                               | 9/**84**/**144**  | 9/**71**/**131** | 9/**49**/**109** |
| maml_omniglot                     | 9/**106**         | 9/**87**         | 9/**43**         |
| resnet152                         | 9/**44**          | 9/**39**         | 9/**24**         |
| resnet18                          | 9/**37**          | 9/**32**         | 9/**17**         |
| resnet50                          | 9/**44**          | 9/**39**         | 9/**24**         |
| functorch_maml_omniglot           | 9/**109**         | 9/**90**         | 9/**45**         |
| pytorch_CycleGAN_and_pix2pix      | 9/**33**          | 9/**29**         | 9/**23**         |
| pytorch_stargan                   | 14/**81**         | 14/**77**        | 14/**36**        |
| pytorch_unet                      | 9/**635**/**635** | 9/**177**/**177**        | 9/**73**/**73**  |
| vgg16                             | 9/**83**          | 9/**69**         | 9/**27**         |

ysiraichi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1719,8 +1744,60 @@ def _update_divisible(self):
+                sympy.Ge(numerator, ((expr.rhs + 1) * denominator))  # type: ignore
+            )
+
+        # The transformations below only work if b is positive.
+        # Note: we only have this information for constants.
+        def is_floordiv_with_positive_denominator(e) -> bool:
+            if not isinstance(e, FloorDiv):
+                return False
+            number = e.args[1]
+            return isinstance(number, sympy.Integer) and bool(number > 0)
+
+        # a // b > expr  => a >= (b + 1) * expr
+        # a // b >= expr => a >= b * expr
+        if isinstance(expr, (sympy.Gt, sympy.Ge)) and is_floordiv_with_positive_denominator(expr.lhs):
+            quotient = expr.rhs if isinstance(expr, sympy.Ge) else (expr.rhs + 1)  # type: ignore
+            expr = sympy.Ge(expr.lhs.args[0], (quotient * expr.lhs.args[1]))  # type: ignore
+        # a // b < expr  => a < b * expr
+        # a // b <= expr => a < (b + 1) * expr
+        if isinstance(expr, (sympy.Lt, sympy.Le)) and is_floordiv_with_positive_denominator(expr.lhs):
+            quotient = expr.rhs if isinstance(expr, sympy.Lt) else (expr.rhs + 1)  # type: ignore
+            expr = sympy.Lt(expr.lhs.args[0], (quotient * expr.lhs.args[1]))  # type: ignore
```
<details>
<summary>Z3 Proof</summary>

```python
tests = [
    (
        # Less-Than
        lambda a,b,c: a / b < c,
        lambda a,b,c: a < c * b,
    ),
    (
        # Less-Than or Equal
        lambda a,b,c: a / b <= c,
        lambda a,b,c: a < (c + 1) * b,
    ),
    (
        # Greater-Than
        lambda a,b,c: a / b > c,
        lambda a,b,c: a >= (c + 1) * b,
    ),
    (
        # Greater-Than or Equal
        lambda a,b,c: a / b >= c,
        lambda a,b,c: a >= c * b,
    ),
]

for old, new in tests:
    a = z3.Int("a")
    b = z3.Int("b")
    c = z3.Int("c")

    solver = z3.Solver()

    solver.add(b > 0)
    solver.add(z3.And(old(a, b, c), z3.Not(new(a, b, c))))

    if solver.check() == z3.sat:
        m = solver.model()
        print(m)
    else:
        print("unsat")
```

</details>

lezcano replied:
Does Z3 integer division agree with the floordiv we're using here?

The semantics of `floordiv` in Z3 are given as per https://smtlib.cs.uiowa.edu/theories-Ints.shtml (and the formulas in the paper in there), by:
![image](https://user-images.githubusercontent.com/3291265/228805914-2893b209-7573-42a2-9375-c021182884fc.png)
This is pretty much an integer division defined so that `0 <= a % b < abs(b)` for every `a,b` irrespective of the sign of `a` and `b`. Note that this is different to Python's division.

Which division are we using when have floordiv in our formulas?

ysiraichi replied:
Ah. You are right. As far as I understand, `FloorDiv` is rounding towards -infinity. I have added an integer division function that hopefully has the same semantics as that.

```python
def div(a, b):
    return (b < 0) * (-a / b) + (b >= 0) * (a / b)
```

ysiraichi replied:
From the link you posted (+ I actually ran `model.evaluate` to check):

> Regardless of sign of m, 
  when n is positive, (div m n) is the floor of the rational number m/n;
  when n is negative, (div m n) is the ceiling of m/n.

ysiraichi replied:
Ah... I missed one negative sign on `b`.

ysiraichi committed:
Update on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2293,6 +2381,95 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
+        # original expression: a < b ==> b > a
+        exprs = [guard.expr]
+
+        if type(guard.expr) in RELOP_MIRROR:
+            exprs.append(RELOP_MIRROR[type(guard.expr)](guard.expr.rhs, guard.expr.lhs))  # type: ignore[arg-type]
```
Rather than doing things twice, why don't you simply flip the expression if needed to leave the free variable at the left?

ysiraichi replied:
I guess that makes sense for univariate expressions. But, for multi-variate expressions, I still think we need this (e.g. `s0 < s1 // 2` -- not a real one). Should I move it to the next PR?

ezyang replied:
I'm fine with it as is

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2087,6 +2164,16 @@ def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
+            rhs_const = get_added_const(expr.rhs)
+            if (
+                    lhs_const is not None
+                    and rhs_const is not None
```
Why does the `rhs_const` need to not be none here?

ysiraichi replied:
Hmm. Good question. Since we want to possibly isolate a symbol on the left-hand side, maybe we could ignore what is in the right-hand side.

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2087,6 +2164,16 @@ def simplify(self, expr: "sympy.Expr") -> "sympy.Expr":
+            if (
+                    lhs_const is not None
+                    and rhs_const is not None
+                    and lhs_const != 0
```
I think we should do this transformation even when the `lhs_const` is zero. In that case, you should just unwrap the inner expression from the lhs and remove the zero.

ysiraichi replied:
I would say we could remove it, since `sympy` simplifies it automatically. But, what if we have something like `0 < expr`?

lezcano replied:
Ok, so I?m a bit lost here. Do we want to do this simplification in other cases as well, or just in the setting of this PR. In the setting of this PR, we cannot have that situation, as we know by construction that we have at least one variable on the LHS.

ysiraichi replied:
On second thoughts, I don't think it makes any significant difference given that `lhs_const == 0`

lezcano reviewed (commented):
Could you please add a few tests as well?

ysiraichi commented:
I don't think these 2 CI failures are related to this PR: some commits after this one (e.g. `f9ca48ddb58196fa2003d6253c38c84567893e34`) also had those very same failures.

ezyang commented:
This is still too complicated. I am not merging it until we have translation validation.

ysiraichi committed:
Update on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix + using sympy_interp. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Temporarily turing translation validation on by default. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix CI issues. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Simplify FX graph creation + CI fixes. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix lint errors. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Add comment. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix CI issues. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Improve python types. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix lint + checking for guard suppression. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Add tests. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Add warning logs temporarily. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3582,6 +3671,94 @@ def guard_from(expr, concrete_val):
 
         return concrete_val
 
+    def refine_ranges(self, guard: ShapeGuard) -> None:
```
It would be helpful to have a top level comment saying what the basic thing this function does is

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3582,6 +3671,94 @@ def guard_from(expr, concrete_val):
+                not isinstance(expr, sympy.Rel)
+                or not isinstance(expr.lhs, sympy.Symbol)
+                or expr.lhs not in self.var_to_range
+            ):
```
nit, we're essentially doing a pattern match here, so to me it would be more clear to demorganize into not (X and Y) to show that afterwards X and Y are true.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3293,6 +3323,64 @@ def _update_divisible(self):
+            Returns an integer constant being added at the top-level of this expression.
+            """
+            if isinstance(expr, sympy.Integer):
+                return expr
```
Hmm, you sure you want this case?

ysiraichi replied:
I don't think so.

ysiraichi committed:
Update on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Revert translation validation by default. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix + comments. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Move Z3 CI dependency to first PR. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Move validator into its own file. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Add test for specialization. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Move tests to #103611. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Handle unknowns correctly. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix success flag on z3.unknown. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Skipping high-latency benchmarks. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Skipping high-latency TIMM benchmarks. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Disable TV for tests that timeout. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Small fixes. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Kill temporary debugging code. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]

ysiraichi committed:
Add Z3 to Windows CI dependency. on "Value range refinement using uni-variate expressions."


This PR introduces value range refinement of shape symbols by symbolically evaluating the
value range of the involved guards. This should help `_maybe_evaluate_static` to eliminate
more guards.

This is a stack of PRs created from the discussion on: #96616.

In summary, this PR:
- simplifies `FloorDiv` nodes on the left-hand side of an expression so as to isolate a
symbol in the numerator
- tries to match the expression against the form: `<symbol> <relop> <expr>`
- uses the matched expression for refining the value range of `<symbol>` using the range
of `<expr>`

[ghstack-poisoned]
